Add ARM64 build

ciaranmcnulty opened the pull request:
I can't use the docker image on my M1 Mac under emulation due to a segfault[^1], so this is a start at multi-architecture build including ARM64 (see #11)

This does not (yet) lead to a successful build of the cucumber project, just a successful `docker build`. I'll work through the rest of the issues and see whether they need to be fixed in the other repo or this one, hence the draft status

 - [x] adds .dockerignore
 - [x] switch to buildx for easier multi-arch 
 - [x] change download URLs that are architecture-specific
 - [x] Javascript builds in monorepo
   - [x] remove Chrome if on ARM (no ARM build is available) 
   - [x] see if Chromium can replace Chrome (only used for html-formatter right now)
 - [x] fix the CI build (Matt broke it, not sure how)
 - [x] Java builds in monorepo (something about `JAVA_HOME` not set?)
 - [x] dotnet builds in monorepo
 - [x] Ruby builds in monorepo
 - [ ] cucumber-js builds
 - [ ] cucumber-jvm builds
 - [ ] cucumber-ruby builds

[^1]: It seems a long route to get this sort of thing fixed in QEMU

ciaranmcnulty committed:
Add a .dockerignore

Prevents the layers being invalidated by e.g. changes to README or the Dockerfile

ciaranmcnulty committed:
Get the image building (without Chrome for now)

mattwynne committed:
Merge branch 'main' into multi-arch

mattwynne committed:
Install chromium-browser binary on arm64 architecture

I finally figured this out from https://askubuntu.com/a/1258911

@ciaranmcnulty I guess you may be able to clean this up, I am not even
sure if the platform guard clauses work!

mattwynne committed:
Use default make target for CI

ciaranmcnulty committed:
Combine some stages

ciaranmcnulty committed:
Remove buildx for now

ciaranmcnulty committed:
Move TARGETARCH before its first usage

ciaranmcnulty committed:
Try and fix the braces around the conditionals

ciaranmcnulty commented:
@mattwynne not sure why the build is failing on Intel - is probably downloading the wrong file(s) for dotnet

mattwynne committed:
Extract dotnet install stuff to a script

Hopefully this will make it a bit easier to see where the bug is

mattwynne committed:
Try handling dotnet version differently

mattwynne committed:
try docker layer caching

mattwynne committed:
Fix use of variables in logging

Still puzzled why TARGETARCH doesn't seem to be set in GitHub Actions

mattwynne committed:
Fail fast

mattwynne committed:
Get order of params right

mattwynne committed:
Try setting DOCKER_BUILDKIT

mattwynne committed:
try to trigger CI

ciaranmcnulty committed:
Merge branch 'main' into multi-arch

mattwynne committed:
Tweak targetarch logic so it doesn't exit 1

mattwynne committed:
Logic was still wrong!

mattwynne committed:
Simplify chrome install

Instead of using one way for intel and one way for arm processors, I
think we can just use nix to install chromium-browser for either
architecture.

mattwynne committed:
Add back version tagging

mattwynne committed:
Merge branch 'main' into multi-arch

mattwynne committed:
Remove duplicate make target and clean up docker push tags

mattwynne committed:
try and get docker-push make target to work

mattwynne committed:
Whoops

mattwynne committed:
Add a make target to build all platforms

mattwynne committed:
Fix build / push so they actually work

mattwynne committed:
Try to work out the right default platform

ciaranmcnulty commented on Makefile:
```diff
@@ -1,17 +1,12 @@
-NAME    := cucumber/cucumber-build
-VERSION := 0.2.0
+NAME      := cucumber/cucumber-build
+VERSION   := 0.2.0
```
Why is this needed? Can we just omit the `—platforms` flag if the env var isn’t set?

mattwynne replied:
I think we need at least the version number checked in somewhere, for publishing the image.

mattwynne committed:
Merge branch 'main' into multi-arch

mattwynne committed:
Build buildx to publish too

mattwynne committed:
Merge branch 'main' into multi-arch

ciaranmcnulty commented on Makefile:
```diff
@@ -1,28 +1,18 @@
-LATEST  := ${NAME}:latest
+NAME      := cucumber/cucumber-build
+VERSION   := 0.2.0
+DEFAULT_PLATFORM = $(shell [ $$(arch) = "arm64" ] && echo "linux/arm64" || echo "linux/amd64")
```
There's actually a build arg for this already that includes the `linux/`

ciaranmcnulty replied:
https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope

mattwynne replied:
aye, but how could you use that here to give people a sensible default when they run `make`?

mattwynne committed:
Merge branch 'main' into multi-arch

ciaranmcnulty commented on Makefile:
```diff
@@ -42,3 +42,7 @@ docker-run-with-secrets: default
 	  -it ${NAME} \
 	  bash
 .PHONY: run-with-secrets
+<<<<<<< HEAD
```
@mattwynne bad merge

mattwynne committed:
Fix bad merge

mattwynne committed:
Remove mono

Going to assume we get success with https://github.com/cucumber/common/issues/1532

mattwynne commented:
I've raised https://github.com/NixOS/nixpkgs/issues/122504 about the errors installing chromium. I'm wondering if we'll have to do without it for now, maybe just ignore these tests in the monorepo somehow.

mattwynne committed:
Add a `local` target for copying the built image to local docker

mattwynne committed:
Replace nix with another chrome download method

Inspired by https://github.com/gidztech/docker-chromium

mattwynne committed:
Be bold and see if we can get GitHub Actions to build both archs

mattwynne committed:
Try to get multi-arch builder working

mattwynne committed:
Try again

mattwynne committed:
Try a more sophisticated GitHub Actions setup

mattwynne committed:
Download specific arch of chrome

It doesn't work yet, I see an error when I actually try to run the
chrome executable.

mattwynne commented:
Today I've tried to script the chrome download so we download the right binary for arm linux (see dfa5cf6)

When I try to run that binary, I see this:

```
cukebot@416564d37d54:/app$ ls -la /usr/bin/chromium-browser 
lrwxrwxrwx 1 root root 33 May 11 16:55 /usr/bin/chromium-browser -> /home/cukebot/chrome-linux/chrome
cukebot@416564d37d54:/app$ /usr/bin/chromium-browser 
/lib/ld-linux-armhf.so.3: No such file or directory
```

So I guess that either:

- I've picked the wrong binary
- We need some other stuff installed for the binary to work
- We need to do something less naive than just simlinking to the main binary program, and add some stuff to `$PATH`
- Something else

Any thoughts @ciaranmcnulty?

mattwynne commented:
OK I *think* this is good to go.

I've tested it in the monorepo and there are still some issues building it on the arm platform, but I think that's now down to assumptions about the architecture in those build scripts, rather than anything wrong with this container.

It would be useful to test the `amd64` platform still works OK though I guess. Maybe I'll do that with a manual build/push and a PR on the monorepo.

mattwynne committed:
Try yet another way of downloading chromium

Seems to actually work!

mattwynne committed:
Docker push commands that actually work.

I guess we could / should use builldx for pushing really

mattwynne commented:
I've manually pushed the container images to [cucumber/cucumber-build:pr-12](https://hub.docker.com/layers/cucumber/cucumber-build/pr-12/images/sha256-60be01ea805535d635211b671aa0bb75415d5792d2151990f18b9f8a4183d14a?context=explore)

mattwynne committed:
Push directly from buildx

mattwynne committed:
Long shot to see if GitHub Actions can build with mono

mattwynne committed:
Try installing mono-runtime instead of mono-devel

mattwynne committed:
Remove elixir install for now; it's failing

mattwynne committed:
Move erlang later to test it

mattwynne committed:
Push a unique tag

mattwynne commented:
OK this is released in v0.4.4 of the container. Huge thanks to @ciaranmcnulty for doing all the hard work to make this happen!

ciaranmcnulty commented:
@mattwynne i think the commits tell a different story ;)

mattwynne commented:
> @mattwynne i think the commits tell a different story ;)

It's quality not quantity that counts! I've learned loads about docker from working with you on this.
